OpenSim/Common/Path.h

```diff
@@ -0,0 +1,162 @@
+
+#include "OpenSim/Common/Object.h"
+
+namespace OpenSim {
+
+//==============================================================================
+//                                 OPENSIM PATH
+//==============================================================================
+/**
+* An abstract class for handling a Path. A Path refers to a list of strings that
+* represent a different level in a hierarchical structure. Each level is divided
+* by designated separators (e.g., "/" or "\"). The string name related to each
+* level is called a "pathElement" here.
+*
+* This class stores the list of levels as a vector of strings. One char is used
+* to denote a path separator when either reading in or writing the Path out to
+* a string. A bool is stored to determine whether this Path should be resolved
+* relative to the root (denoted with a leading separator char) or not.
+*
+* @author Carmichael Ong
+*/
+class OSIMCOMMON_API Path {
+
+public:
+    /// Default constructor
+    Path() = default;
+
+    /// Construct Path from a string, given separator character and a string
+    /// of invalidChars. Performs a cleanPath() at the end of construction.
+    Path(const std::string path,
+         const char separator,
+         const std::string invalidChars);
+
+    /// Construct Path from a vector of strings (pathVec), given separator
+    /// character and a string of invalidChars. Performs a cleanPath() at 
+    /// the end of construction.
+    Path(std::vector<std::string> pathVec,
+         const char separator,
+         const std::string invalidChars,
+         bool isAbsolute);
+
+    /// Use default copy constructor and assignment operator.
+    Path(const Path&) = default;
+    Path& operator=(const Path&) = default;
+
+    /// Destructor.
+    ~Path() = default;
+
+    /// Write out the path to a string with each element separated by the
+    /// specified separator.
+    std::string getString() const;
+
+    /// Return true if this Path is an absolute path.
+    bool isAbsolute() { return _isAbsolute; };
+
+    /// Return the number of elements in the Path
+    size_t getPathLength() const { return _path.size(); };
+
+    /// Push a string to the back of a path (i.e. the end). First checks if the
+    /// pathElement is legal. Does not change if path is absolute or relative.
+    void pushBack(const std::string& pathElement) {
+        appendPathElement(pathElement);
+    }
+
+    /// Pure virtual function that returns a char for the designated separator.
+    virtual const char getSeparator() const = 0;
+
+    /// Pure virtual function that returns a string of invalid characters.
+    virtual const std::string getInvalidChars() const = 0;
+
+protected:
+    /// Get an absolute path by resolving it relative to a given otherPath.
+    /// If the current Path is already absolute, return the same Path.
+    std::vector<std::string> formAbsolutePathVec(Path* otherPath);
```

Comment:
I think the parameter should be `const Path& otherPath`. Also, this function should be const. Similar for `formRelativePathVec`.